packages/gatsby/src/schema/node-model.js

```diff
@@ -383,11 +383,10 @@ class LocalNodeModel {
           queryFields,
           actualFieldsToResolve
         )
-        const mergedResolved = _.merge(
-          node.__gatsby_resolved || {},
-          resolvedFields
-        )
-        return mergedResolved
+        if (!node.__gatsby_resolved) {
+          node.__gatsby_resolved = {}
```

Comment:
Should this not actually run the resolvers? The other part (in redux/nodes.ts) does something like this:

```js
        if (!node.__gatsby_resolved) {
          const typeName = node.internal.type
          const resolvedNodes = resolvedNodesCache.get(typeName)
          const resolved = resolvedNodes?.get(node.id)
          if (resolved !== undefined) {
            node.__gatsby_resolved = resolved
          }
        }
```

Why would we not want to do that here?

Reply:
We actually want. But the problem is that this happens as a part of this call:

https://github.com/gatsbyjs/gatsby/blob/b34680487cff802ed36f7c328da8542e16cbda9d/packages/gatsby/src/redux/run-fast-filters.ts#L232-L240

It will populate the cache for the filter AND will set `__gatsby_resolved`. But in our case first query with this filter has no materialized fields, hence `const resolved = resolvedNodes?.get(node.id)` returns undefined.

Now the second query bypasses this codepath (because we have `filterCacheKey` now) but hits this codepath:

https://github.com/gatsbyjs/gatsby/blob/b34680487cff802ed36f7c328da8542e16cbda9d/packages/gatsby/src/redux/run-fast-filters.ts#L242-L247

And in this codepath we **don't** set `__gatsby_resolved` anywhere. It kinda works in most cases only because we re-assign `__gatsby_resolved` here:

https://github.com/gatsbyjs/gatsby/blob/b34680487cff802ed36f7c328da8542e16cbda9d/packages/gatsby/src/schema/node-model.js#L386-L389

`merge` mutates the first argument with the result of the merge. But on the first run this first argument of `merge` is just a detached object, so it is not mutated and cached results do not get their `__gatsby_resolved` values.

Probably the correct fix would be to add the piece you've highlighted somewhere inside `getNodesFromCacheByValue` but that will be a much bigger change and also will likely require an additional loop over all found nodes.

I went with an `easy` route because it is also consistent - we already mutate `node.__gatsby_resolved` in all cases but the very first one.

Reply:
So the `tldr;` is that when `filtersCache` is populated we bypass this codepath:

```js
        if (!node.__gatsby_resolved) {
          const typeName = node.internal.type
          const resolvedNodes = resolvedNodesCache.get(typeName)
          const resolved = resolvedNodes?.get(node.id)
          if (resolved !== undefined) {
            node.__gatsby_resolved = resolved
          }
        }
```

Reply:
A fun fact: if I change this:

https://github.com/gatsbyjs/gatsby/blob/b34680487cff802ed36f7c328da8542e16cbda9d/packages/gatsby/src/schema/node-model.js#L386-L389

to this:

```js
const mergedResolved = _.merge({}, node.__gatsby_resolved ?? {}, resolvedFields)
```

a bunch of tests fails. So we kinda already rely on this mutation for materialization to function properly.

Reply:
ok :shrug:  :)